docs/docs/icicle/programmers_guide/go.md

```diff
@@ -1 +1,289 @@
+
+Retrieve the number of available devices and check if a pointer is allocated on the host or on the active device:
+
+```go
+numDevices := runtime.GetDeviceCount()
+
+var ptr unsafe.Pointer
+isHostMemory = runtime.IsHostMemory(ptr)
+isDeviceMemory = runtime.IsActiveDeviceMemory(ptr)
+```
+
+## Memory Management
+
+### Allocating and Freeing Memory
+
+Memory can be allocated and freed on the active device:
+
+```go
+ptr, err := runtime.Malloc(1024) // Allocate 1024 bytes
+err := runtime.Free(ptr) // Free the allocated memory
+```
+
+### Asynchronous Memory Operations
+
+You can perform memory allocation and deallocation asynchronously using streams:
+
+```go
+stream, err := runtime.CreateStream()
+
+ptr, err := runtime.MallocAsync(1024, stream)
+err = runtime.FreeAsync(ptr, stream)
+```
+
+### Querying Available Memory
+
+Retrieve the total and available memory on the active device:
+
+```go
+size_t total_memory, available_memory;
+availableMemory, err := runtime.GetAvailableMemory()
+freeMemory := availableMemory.Free
+totalMemory := availableMemory.Total
+```
+
+### Setting Memory Values
+
+Set memory to a specific value on the active device, synchronously or asynchronously:
+
+```go
+err := runtime.Memset(ptr, 0, 1024) // Set 1024 bytes to 0
+err := runtime.MemsetAsync(ptr, 0, 1024, stream)
+```
+
+## Data Transfer
+
+### Explicit Data Transfers
+
+To avoid device-inference overhead, use explicit copy functions:
+
+```go
+result := runtime.CopyToHost(host_dst, device_src, size)
+result := runtime.CopyToHostAsync(host_dst, device_src, size, stream)
+result := runtime.CopyToDevice(device_dst, host_src, size)
+result := runtime.CopyToDeviceAsync(device_dst, host_src, size, stream)
+```
+
+## Stream Management
+
+### Creating and Destroying Streams
+
+Streams are used to manage asynchronous operations:
+
+```go
+stream, err := runtime.CreateStream()
+err = runtime.DestroyStream(stream)
+```
+
+## Synchronization
+
+### Synchronizing Streams and Devices
+
+Ensure all previous operations on a stream or device are completed before proceeding:
+
+```go
+err := runtime.StreamSynchronize(stream)
+err := runtime.DeviceSynchronize()
+```
+
+## Device Properties
+
+### Checking Device Availability
+
+Check if a device is available and retrieve a list of registered devices:
+
+```go
+dev := runtime.CreateDevice("CPU", 0)
+isCPUAvail := runtime.IsDeviceAvailable(dev)
+```
+
+### Querying Device Properties
+
+Retrieve properties of the active device:
+
+```go
+properties, err := runtime.GetDeviceProperties(properties);
+
+/******************/
+// where DeviceProperties is
+type DeviceProperties struct {
+  UsingHostMemory      bool   // Indicates if the device uses host memory
+  NumMemoryRegions     int32  // Number of memory regions available on the device
+  SupportsPinnedMemory bool   // Indicates if the device supports pinned memory
+}
+```
+
+
+TODO
+
+
+
+
+## Compute APIs
+
+### Multi-Scalar Multiplication (MSM) Example
```

Comment:
Why are the examples here in cpp?

Reply:
The layout was taken from cpp guide and this portion hasn't been update for Go yet